README.md
[![npm](https://img.shields.io/npm/v/happn-cluster.svg)](https://www.npmjs.com/package/happn-cluster)
[![Build Status](https://travis-ci.org/happner/happn-cluster.svg?branch=master)](https://travis-ci.org/happner/happn-cluster)
[![Coverage Status](https://coveralls.io/repos/happner/happn-cluster/badge.svg?branch=master&service=github)](https://coveralls.io/github/happner/happn-cluster?branch=master)

# happn-cluster

Extends happn with cluster ability.

Requires that each cluster member mounts the same shared data service. See [happn-service-mongo](https://github.com/happner/happn-service-mongo).

See also [happn cluster aws example](https://github.com/happner/happn-cluster-aws-example)

## Install

`npm install happn-cluster happn-service-mongo-2 --save`

Note data service installed separately.

## Configure

See [happn](https://github.com/happner/happn) for full complement of happn config.

```javascript
var HappnCluster = require('happn-cluster');
var defaultConfig = {
  // name: undefined,  // defaults from happn service host:port (10-0-0-1_55000)
  // host: '0.0.0.0', // happn service ip
  // port: 57000,    // happn service port
  // secure: true,  // to enable security
  // announceHost: 'externally-visible-ip-or-hostname', // eg. when docker
  services: {

    // // security sub-config (to enable security)
    // security: {
    //   config: {
    //     adminUser: {
    //       username: '_ADMIN', // <---- leave this as _ADMIN
    //       password: 'happn'
    //     }
    //   }
    // },

    // shared data plugin sub-config (defaults displayed)
    data: {
      config: {
        datastores: [
          {
            name: 'mongo',
            provider: 'happn-service-mongo-2',
            isDefault: true,
            settings: {
              collection: 'happn-cluster',
              database: 'happn-cluster',
              url: 'mongodb://127.0.0.1:27017'
              // url: 'mongodb://username:password@127.0.0.1:27017,127.0.0.1:27018,127.0.0.1:27019/happn?replicaSet=test-set&ssl=true&authSource=admin'
            }
          }
        ]
      }
    },

    // proxy sub-config (defaults displayed)
    proxy: {
      config: {
        port: 55000,
        host: '0.0.0.0',
        allowSelfSignedCerts: false,
        // timeout: 20 * 60 * 1000, // request socket idle timeout
        // defer: false,
        // keyPath: 'path/to/key',
        // certPath: 'path/to/cert'
      }
    },

    // orchestrator sub-config (defaults displayed)
    orchestrator: {
      config: {
        minimumPeers: 1,
        replicate: ['*'],
        stableReportInterval: 5000,
        stabiliseTimeout: 120 * 1000 // 0 disables
      }
    },

    // membership sub-config (defaults displayed)
    membership: {
      config: {
        clusterName: 'happn-cluster',
        seed: false,
        seedWait: 0,
        randomWait: 0,
        // host: undefined, // defaults to first public IPv4 address
        port: 56000,
        // hosts: [],
        joinTimeout: 2000,
        pingInterval: 1000,
        pingTimeout: 200,
        pingReqTimeout: 600,
        pingReqGroupSize: 3,
        udp: {
          maxDgramSize: 512
        }
        disseminationFactor: 15
      }
    },
    replicator: {
      config: {
        //when users, groups or permissions change - the security directory updates are 
        //batched and pushed out to the cluster, every 3000ms - you can change this to make
        //the pushes happen less or more often, the batches are also de-duplicated by 
        //whatHappend (link-group, unlink-group etc) and the user or group name in question.
        securityChangesetReplicateInterval: 3000 //once every 3 seconds
      }
    }
  }
}

HappnCluster.create(defaultConfig)
  .then(function(server) {
    // ...
  })
  .catch(function(error) {
    process.exit(0);
  });

```

## NB: by default replicate will replicate all events across the cluster (see above config.replcate is ['\*']) - because we are now using happn-3 version 8.0.0 the / and \* characters have stricter rules, ie: to replicate all events for a set or remove with path /my/test/event you need to set replicate to ['/my/\*/\*'], /my/\* and /my\* or /my/te\*/event will no longer work.

### Happn Config

#### name

Each happn node in the cluster requires a unique name. The name will default from the host:port below to produce something like '10-0-0-1_55000'. If happn is configured to listen at 0.0.0.0 then the name will default instead from the first public IPv4 address found in interfaces.

#### host

The host/ip that the happn service should listen on. Also supports [dface](https://github.com/happner/dface) spec for cloning nodes into a DHCP environment.

#### port

The port that the happn service should listen on.

#### secure

Set true to enable security.

**If one cluster node is secure, they all need to be secure.**

#### announceHost

When using docker the internal container ip is announced into the cluster for remote members to connect to. But that ip is not accessible, use this config to specify the ip or hostname of the accessible external interface. Note that a similar issue occurs with the ports so all ports require a **same-to-same mapping when using docker**.

### Security Sub-Config

Inter-cluster connections are made using the admin user. All nodes in the cluster will need the same
admin username and password configured.

**Once created, the admin user's password cannot be changed from config.**

To change the admin password.

* Stop all cluster nodes.
* Put the new password into all cluster node configs.
* Delete the old \_ADMIN user and \_ADMIN group membership from the shared database.

```bash
mongo mongodb://127.0.0.1:27017/happn-cluster
> use happn-cluster
> db['happn-cluster'].remove({path: {$in: ['/_SYSTEM/_SECURITY/_USER/_ADMIN', '/_SYSTEM/_SECURITY/_USER/_ADMIN/_USER_GROUP/_ADMIN']}});
```
* Restart the cluster (admin user will be regenerated with new password)

**The above also applies after starting a cluster with security unconfigured. The admin user is still created with the default password 'happn'. Upon turning on security later the password will then need to be changed as described above.**

### Shared Data Sub-Config

By configuring a shared data service all nodes in the cluster can serve the same data to clients. The
default uses the [happn mongo plugin](https://github.com/happner/happn-service-mongo). The localhost
url is porbably not what you want.

### Proxy Sub-Config

A starting cluster node immediately starts the happn service listening to allow the orchestrator to
establish the inter-cluster replication bridges.

Clients attaching to this happn service port will therefore be connecting before the node is ready.

Instead, clients should connect through the proxy port, whose start is pended until the node is ready.

#### config.[port, host]

The socket address where the proxy listens for clients.

#### config.[keyPath, certPath]

Specify key and cert filenames to switch switch on https

#### config.allowSelfSignedCerts

Allow connecting to local happn which is listening with https and self signed cert.

#### config.defer

Set to true to not start the proxy service. If set the `services.proxy.start()` will need to be called externally.

#### config.timeout

Configure proxy's request-side socket's idle timeout. Default 20 minutes.

### Orchestrator Sub-Config

#### config.minimumPeers

This pends the starting of the proxy until there are this many known peers in the cluster. This prevents
the `thundering herd` (of clients) from settling all their sockets permanently onto the first started node.

#### config.replicate

Array of happn paths or path masks that will be replicated throughout the cluster.

#### config.stableReportInterval

Having received the membership list (other cluster nodes), the orchestrator stalls the startup
procedure (pending the proxy start) until fully connected (stabilised). This interval controls
the frequency with which the outstanding connection states are reported into the log.

#### config.stabiliseTimeout

Defines how long to wait for this starting node to become fully connected (stabilised) before giving up and stopping the node. In all known cases a starting node will either reach stability or fail explicitly with an error. This is a failsafe (for unknown cases) to prevent endlessly awaiting stability where it would be better to stop and try joining the cluster again.

**Note that this acts in opposition to `minimumPeers` - A starting node awaiting minimum peers will still time out.**

### Membership Sub-Config

#### config.clusterName

Every member of the cluster should have the same configured `clusterName`.
The name is limited to characters acceptable in happn paths, namely '\_*-', numbers and letters.
Joining members with a different clusterName will be ignored by the orchestrator.

#### config.seed

Boolean flag sets this member as the cluster seed member. If `true` this member will not terminate
upon failing to join any other cluster members and can therefore enter a cluster successfully as
the first member.

Each other member should include the seed member, among others, in their **config.hosts**
list of hosts to join when starting.

#### config.seedWait

Members that are not the seed member pause this long before starting. This allows for a booting host that
contains multiple cluster member instances all starting concurrently where one is the seed member. By waiting, the seed member will be up and running before the others attempt to join it in the cluster.

#### config.randomWait

Members that are not the seed member pause a random amount of milliseconds from this range before starting. This is to alleviate the problem of SWIM not handling large numbers of concurrent cluster joins.

Be sure to set this value to less than the stabiliseTimeout.

#### config.[host, port]

The host and port on which this member's SWIM service should listen. Host should be an actual ip address
or hostname, not '0.0.0.0'. It can also be specified using [dface](https://github.com/happner/dface) spec.

**Important: The membership service protocol is currently insecure. These ports need to be protected in a private cluster.** [issues/1](https://github.com/happner/happn-cluster/issues/1)

#### config.hosts

The list of initial cluster members via which this member joins the cluster. This should include the
seed member and a selection of other members likely to be online.

Items in the list are composed of `host:port` as configured on the remote members' **config.host**
and **config.port** above.

Example: `['10.0.0.1:56000', '10.0.0.2:56000', '10.0.0.3:56000']`

It is **strongly recommended** that all nodes in the cluster use the same **config.hosts** list to avoid
the posibility of orphaned subclusters arising. It must therefore also be ensured that at least one of
the hosts in the list is online at all times. They can be upgraded one at a time but not all together.

In the event of all nodes in the **config.hosts** going down simultaneously the remaining nodes in the
cluster will be orphaned and require a restart.

#### config.joinTimeout

The bootstrapping SWIM member waits this long to accumulate the full membership list from the network.

#### config.pingInterval

The running SWIM member `cycles` through it's member list sending a ping to determine if the member is still there. A ping is sent once every interval. The default 1000ms results in a noticable delay in detecting departed members. It's a tradeoff between cluster-size/detection-time/ping-bandwidth. Keep in mind that all members are doing the cyclic ping so worst-case discovery time is not `1000ms * memberCount`.

#### config.pingTimeout

The running SWIM member expects a reply to its ping. Receiving none within this time results in the pinged member coming under suspicion of being faulty/offline. At this point secondary ping requests are sent to a random selection of other members to ping the suspect themselves to confirm the suspicion.

#### config.pingReqTimeout

The running SWIM member expects a reply from those secondary ping requests within this time. If not received the suspect is declared faulty/offline and this information is disseminated into the cluster.

#### config.pingReqGroupSize

Secondary ping requests are sent to this many other members.

#### config.[udp, disseminationFactor]

Members updates (arrived/departed) are disseminated throughout the cluster on the back of the pings already being sent. **udp.maxDgramSize** limits the size of those payloads. **disseminationFactor** combined logarithmically with cluster size determines for how many pings in the `cycle` any given membership update remains eligible for dissemination. All eligible updates are sent with every ping up to the available **maxDgramSize**.

See [swim.js](https://github.com/happner/swim-js)

## Deployment orchestration and automation

Details of automating happn-cluster using Docker and Ansible can be found here: [https://github.com/happner/happner-ansible-orchestration](https://github.com/happner/happner-ansible-orchestration).


RELEASES.md
2.5.5 2017-03-21
----------------
  - activated primus destroy bug dodging

2.5.6 2017-03-21
----------------
  - activated primus destroy bug dodging

2.5.7 2017-03-23
----------------
  - onward release happn-3

2.6.0 2017-04-06
----------------
  - onward release happn-3

2.6.1 2017-04-18
----------------
  - onward release of happn-3

2.7.0 2017-05-20
----------------
  - added `config.services.proxy.config.defer` to not start proxy for use with happner-cluster

3.0.0 2017-08-19
----------------
  - update to happn-3 ^3.0.0
  - test node version 8

4.0.0 2017-11-08
----------------
  - update to happn-3 ^5.0.0

4.1.0 2017-11-25
----------------
  - update to happn-3 ^5.1.1
  - add happn-stats

non-release 2017-12-14
----------------------
  - expanded 19-restart-peer test to more aggressively test stabilize
  - removed benchmarket

4.2.0 2018-01-09
----------------
  - added configurable cluster proxy socket idle timeout

4.3.0 2018-03-27
----------------
  - add replication service

5.0.0 2018-04-12
----------------
  - onward release happn-3

6.0.0 2018-05-24
----------------
  - onward release happn-3

6.0.1 2018-06-28 - unpublished
------------------------------
  - issue #36 - replace http-proxy with bouncy. Enables handling non-standard binary payloads.

6.0.2 2018-06-29
----------------
  - rolling back bouncy update

6.1.0 2018-06-29
----------------
  - persistMembers membership service change

7.0.0 2018-19-10
----------------
  - happn-3 version 8.0.0

7.0.1 2018-09-16
----------------
  - happn-3 version 8.0.1

7.0.2 2018-10-27
----------------
  - happn-3 version 8.0.3

7.1.0 2018-10-27
----------------
  - happn-3 version 8.1.1

7.2.0 2018-11-17
----------------
  - happn-3 version 8.2.1

7.2.1 2018-11-17
----------------
  - happn-3 version 8.2.7

8.0.0 2019-03-01
----------------
  - happn-3 version 9.0.0

8.0.1 2019-03-01
----------------
  - happn-3 version 9.0.1

8.0.2 2019-03-01
----------------
  - happn-3 version 9.0.2

8.0.3 2019-04-26
----------------
  - happn-3 version 9.0.4
  - extended tests around cluster member persistence

8.1.0 2019-04-26
----------------
  - happn-3 version 10.0.0

8.1.1 2019-04-26
----------------
  - ws 7 fix

8.1.2 2019-04-26
----------------
  - ws error bubble fix
  - happner and happn upgrades

8.1.3 2019-08-21
----------------
  - issue with stable timeout not being cleared on stop

8.1.4 2019-09-27
----------------
  - updated happn-3 dependency

9.0.0 2019-11-28
----------------
  - happn-3 version 11.0.0, breaking: client disconnection on token revocation

9.0.1 2019-01-25
----------------
  - linting

9.0.2 2020-02-24
----------------
  - happn-3 version 11.2.0, zombie socket cleanup

9.0.3 2020-02-24
----------------
  - fix: member client will not double disconnect if server is stopped more than once
  - cleanup - removed dead code and comment from orchestrator

9.0.4 2020-06-24
----------------
  - dep: bumped happn-3 version 11.5.2
  - dep: changed sillyname dependency to happn-sillyname

9.0.5 2020-07-07
----------------
  - Added ability to set NETWORK_INTERFACE for utils/get-address

9.0.6 2020-10-04
----------------
  - happn-3 patch: selective security cache clearing and concurrency 1 queue on dataChanged event - SMC-1189
  - patch: SMC-1252 - cluster member restarts cause leak of peer admin

10.0.0 2020-11-04
-----------------
  - happn-cluster upgrade - SMC-1356 - batching of security updates

10.1.0 2020-11-04
-----------------
  - SMC-1285 - periodic health state logging
  - SMC-1284 - re-establishment of cluster member on socket failure

10.1.1 2021-03-31
-----------------
  - SMC-1808 - user permissions dependencies updated

10.2.0 2021-07-20
-----------------
  - SMC-1810 - nested permissions for set and on

10.3.0 2021-08-04
-----------------
  - SMC-3611 - getAddress now a closure with logger passed in, also allows nic id and address index to be defined - will resolve to an address with a warning if nic or address index does not match an existing item
  - SMC-3646 - set up coveralls and github actions

10.3.1 2021-08-27
------------------
- SMC-3242: happn-3 update fixed onBehalfOf and nested permissions issue

10.3.2 2021-09-11
------------------
- SMC-4044 - cookie events not dependent on client login

10.3.3 2021-09-20
------------------
- SMC-4146 - happn-3 upgrade 11.13.4

10.3.4 2021-09-28
-----------------
  - happn-3 upgrades:
  - fix: SMC-4209 - concurrency issue, user created logged on deleted, causes security directory update to fatal
  - fix: SMC-4208 - merge insert now uses upsert, moved constants out of data service

11.0.0 2021-10-07
-----------------
  - SMC-3823: mongo version 4 driver,
  - SMC-4309: removed persistMembers functionality

11.0.1 2021-11-11
-----------------
  - fix: SMC-4512 - happn-3 upgrade, db error causes fatal

11.1.0 2021-11-24
-----------------
  - happn-3 upgrade:
  - feature: SMC-2954 - Allows for configuration and use of multiple authentication providers.
  - fix: SMC-4386 -  upsertMultiplePermissions will now allow for removing permissions/prohibitions as well as upserting permissions or prohibitions

11.2.0 2021-12-02
-----------------
  - SMC-734: removed bitcore-lib and encrypted payloads

11.2.1 2021-12-04
-----------------
  - SMC-4466: removed bitcore-lib and encrypted payloads, update of happn-3

11.3.0 2022-01-04
-----------------
  - SMC-4550: updated happn-3 dependency

11.3.1 2022-01-10
-----------------
  - SMC-4938: lookup tables fail for rest request
  - SMC-4550: lookup tables security sync items

11.3.2 2022-01-19
-----------------
  - SMC-4550: Added lookup table SD events to unbatchSecurityUpdate in replicator

11.3.3 2022-01-19
-----------------
  - SMC-4550: bumped happn-3 versions

lib/membership.js
module.exports = Membership;

var EventEmitter = require('events').EventEmitter;
var util = require('util');
var Swim = require('happn-swim');
var Promise = require('bluebird');
var dface = require('dface');

var property = require('./property');
var getAddress = require('./get-address');

function Membership(opts) {
  property(this, 'log', opts.logger.createLogger('Membership'));
}

util.inherits(Membership, EventEmitter);

Membership.prototype.initialize = function(config, callback) {
  property(this, 'happn', this.happn);
  property(this, 'config', config);
  // this.config = config;
  this.members = {};
  this.__defaults(callback);
};

Membership.prototype.stop = function(options, callback) {
  if (typeof options == 'function') callback = options;
  this.log.info('stopping');

  if (this.swim) {
    this.swim.leave();
    this.swim.removeAllListeners();
  }

  this.log.debug('stopped');
  callback();
};

Membership.prototype.bootstrap = Promise.promisify(function(callback) {
  var config, protocol, address, happnUrl, wait, _this = this;

  this.log.info('listening at %s', this.swimAddress);
  this.log.info('joining cluster \'%s\'', this.config.clusterName);

  config = this.config;
  protocol = this.happn.config.transport.mode;
  address = this.happn.server.address();
  if (address.address == '0.0.0.0') {
    // using this to inform remote hosts where to attach
    // 0.0.0.0 won't do, instead use first public ipv4 address
    happnUrl = getAddress();
  } else {
    happnUrl = address.address;
  }
  happnUrl += ':' + address.port;
  happnUrl = protocol + '://' + happnUrl;

  if (this.swim) {
    this.swim.leave();
    this.swim.removeAllListeners();
  }

  property(this, 'swim', new Swim({
    local: {
      host: this.swimAddress,
      meta: {
        cluster: config.clusterName,
        url: happnUrl
      }
    },
    disseminationFactor: config.disseminationFactor,
    interval: config.pingInterval,
    joinTimeout: config.joinTimeout,
    pingTimeout: config.pingTimeout,
    pingReqTimeout: config.pingReqTimeout,
    pingReqGroupSize: config.pingReqGroupSize,
    udp: config.udp
  }));

  wait = config.seed ? 0 : config.seedWait;

  setTimeout(function() {
    _this.swim.bootstrap(_this.config.hosts, function(error) {
      if (error) {
        if (error.name == 'JoinFailedError' && !_this.config.seed) {
          return callback(error);
          // seed member accepts failure to join cluster
        }
      }

      var doLog = false;

      function logMemberChange() {
        if (!doLog) return;
        _this.log.info('has %d other members', Object.keys(_this.members).length);
      }

      function addMember(member) {
        if (_this.members[member.host]) return;
        _this.members[member.host] = {
          url: member.meta.url
        };
        logMemberChange();
        _this.emit('add', {
          memberId: member.host,
          url: member.meta.url
        });
      }

      function removeMember(member) {
        if (!_this.members[member.host]) return;
        delete _this.members[member.host];
        logMemberChange();
        _this.emit('remove', {
          memberId: member.host
        });
      }

      function onUpdate(member) {
        if (member.host == _this.swimAddress) return;
        if (member.meta.cluster !== config.clusterName) return;
        if (member.state == 0) return addMember(member);
        if (member.state == 1) return;
        if (member.state == 2) return removeMember(member);
      }

      var members = _this.swim.members();
      _this.log.info('has %d other members', members.length);

      members.forEach(addMember);

      doLog = true;

      _this.swim.on(Swim.EventType.Update, onUpdate);
      _this.swim.on(Swim.EventType.Change, onUpdate);

      callback();
    });
  }, wait);
});

Membership.prototype.__defaults = function(callback) {
  var dataPlugin, config = this.config;

  try {
    // cluster requires shared database (for now)
    // the only shared database plugin is happn-service-mongo
    // (missing some way to test isDatabasePluginShared)
    dataPlugin = this.happn.config.services.data.path;
    if (dataPlugin !== 'happn-service-mongo') {
      return callback(new Error('cluster requires shared data service'));
    }
  } catch (e) {
    return callback(new Error('cluster requires shared data service'));
  }

  if (typeof config.seed !== 'boolean') {
    config.seed = false;
  }

  if (typeof config.seedWait !== 'number') {
    config.seedWait = 0;
  }

  config.joinType = config.joinType || 'dynamic';

  if (config.joinType !== 'static' && config.joinType !== 'dynamic') {
    return callback(new Error('invalid membership joinType'));
  }

  try {
    config.host = dface(config.host);
  } catch (e) {
    return callback(e);
  }

  // need actual ip address, remote hosts can't connect to 0.0.0.0 here
  if (config.host == '0.0.0.0') {
    config.host = getAddress();
  }

  if (!config.port) {
    config.port = 56000;
  }

  if (config.joinType == 'static' && (!config.hosts || config.hosts.length == 0)) {
    return callback(new Error('missing membership.hosts to join'));
  }

  if (typeof config.joinTimeout == 'undefined') config.joinTimeout = 2000;
  if (typeof config.pingInterval == 'undefined') config.pingInterval = 1000;
  if (typeof config.pingTimeout == 'undefined') config.pingTimeout = 200;
  if (typeof config.pingReqTimeout == 'undefined') config.pingReqTimeout = 600;
  if (typeof config.pingReqGroupSize == 'undefined') config.pingReqGroupSize = 3;
  if (typeof config.udp == 'undefined') config.udp = {};
  if (typeof config.udp.maxDgramSize == 'undefined') config.udp.maxDgramSize = 512;
  if (typeof config.disseminationFactor == 'undefined') config.disseminationFactor = 15;

  property(this, 'swimAddress', config.host + ':' + config.port);
  this.membershipId = this.swimAddress;

  callback();
};
